### PR TITLE
Update Kokkos library in LAMMPS to v4.1.0

### PR DESCRIPTION
**Pull request by stanmoore1**
**Summary**

Update Kokkos library in LAMMPS to v4.1.0
NOTE: Requires C++17

**Related Issue(s)**

None

**Author(s)**

Stan Moore (SNL), Kokkos developers

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

Yes

### CODE REVIEW

**Comment by stanmoore1** [comment]
Note: here is a list of minimum compiler versions for Kokkos 4.0: https://github.com/kokkos/kokkos/issues/5285#issue-1320161352

**Comment by stanmoore1** [comment]
Kokkos regression tests passed. Need to check performance.

**Review by rbberger — APPROVED** [review]
The CMake changes look fine to me.

File: cmake/CMakeLists.txt
Comment (junghans) [inline cmake/CMakeLists.txt:147]: I think it would be better to do this on per-target basis with `set_target_properties()`, just in case somebody would like to compile lammps with C++20 or newer.

Comment (akohlmey) [reply cmake/CMakeLists.txt:147]: We've had something like that before, but that led to all kinds of unexpected inconsistencies.
If you want C++-20 you have to use it for "everything" by setting `-DCMAKE_CXX_STANDARD=20`.

Comment (junghans) [reply cmake/CMakeLists.txt:147]: see https://stackoverflow.com/questions/70566491/how-to-bump-the-c-standard-from-the-cmake-command-line

also Kokkos developer @dalg24 had some strong opinions on that.

Comment (dalg24) [reply cmake/CMakeLists.txt:147]: > If you want C++-20 you have to use it for "everything" by setting `-DCMAKE_CXX_STANDARD=20`.

On board with that approach.

Only unfortunate thing I see with the snippet above is if one explicitly configure with `-DCMAKE_CXX_STANDARD=11` or `14` it would silently be changed to `17` without even raising a warning.
You could do something like
```CMake
if(NOT CMAKE_CXX_STANDARD)
  if(PKG_KOKKOS)
    set(CMAKE_CXX_STANDARD 17)
  else()
    set(CMAKE_CXX_STANDARD 11)
  endif()
endif()

if(PKG_KOKKOS)
  if(CMAKE_CXX_STANDARD LESS 17)
    message(FATAL_ERROR "C++ standard must be set to at least 17")
  endif()
else()
  if(CMAKE_CXX_STANDARD LESS 11)
    message(FATAL_ERROR "C++ standard must be set to at least 11")
  endif()
endif()
```

Comment (akohlmey) [reply cmake/CMakeLists.txt:147]: That stackoverflow issue is setting the C++ standard unconditionally. The quoted cmake script code from LAMMPS does not. It will *raise* the C++ standard to 14 (old) or 17 (new) is it has not been set or set to a lower value.

Comment (akohlmey) [reply cmake/CMakeLists.txt:147]: @dalg24 True, but for now I am very happy with a solution that requires to write and maintain only 2 lines and wait until somebody runs into a situation where it creates a problem before considering a more complex approach.

**Review by junghans — APPROVED** [review]

**Comment by opoplawski** [comment]
Any progress here?  Fedora Rawhide has updated kokkos to 4.0 and now the lammps builds are failing.  A new release would be very helpful.  Thanks.

**Comment by stanmoore1** [comment]
@opoplawski I think #3701 will address the issue you are seeing, can you confirm?

**Comment by stanmoore1** [comment]
@akohlmey it looks like the run test failure is unrelated to this PR?

**Comment by stanmoore1** [comment]
Since this PR requires c++17, we are planning to wait until after the next stable release to merge this, to maximize backwards compatibility. Deprecated code was removed in #3701, which should allow LAMMPS to be compatible with an external Kokkos v4.0.

**Comment by akohlmey** [comment]
> @akohlmey it looks like the run test failure is unrelated to this PR?

That is correct. It is a ripple from the removal of installing the python module from "make install". The run and regression tests depend on it, and the replacement has a race condition when running "make install-python" concurrently from the same source checkout. It took me a while to narrow down where the exact issue is since there were multiple problems that needed changes in multiple places.

**Comment by skyreflectedinmirrors** [comment]
FWIW Stan, I still see a bunch of hash map errors:

```
lammps/src/KOKKOS/neigh_bond_kokkos.cpp:284:89: error: reference to __host__ function 'dual_hash_type' in __host__ __device__ function
    int atom1 = AtomKokkos::map_kokkos<DeviceType>(bond_atom(i,m),map_style,k_map_array,k_map_hash);
                                                                                        ^
lammps/src/KOKKOS/neigh_bond_kokkos.cpp:1230:16: note: in instantiation of member function 'LAMMPS_NS::NeighBondKokkos<Kokkos::HIP>::operator()' requested here
template class NeighBondKokkos<LMPDeviceType>;
               ^
lammps/src/KOKKOS/kokkos_type.h:511:8: note: 'dual_hash_type' declared here
struct dual_hash_type {
       ^
lammps/src/KOKKOS/neigh_bond_kokkos.cpp:372:89: error: reference to __host__ function 'dual_hash_type' in __host__ __device__ function
    int atom1 = AtomKokkos::map_kokkos<DeviceType>(bond_atom(i,m),map_style,k_map_array,k_map_hash);
```

when building w/ your branch + Kokkos develop.  I _think_ it's because there's still an implicit constructor in the dual_hash_type:

image

that isn't deleted (and therefore defaults to `__host__`)?

This might be something specific to develop however.

**Comment by stanmoore1** [comment]
@arghdos thanks will take a look.

**Comment by stanmoore1** [comment]
@arghdos I can reproduce the compile errors with Kokkos `develop`, does not seem to apply to 4.0.

**Comment by rbberger** [comment]
@stanmoore1 created https://github.com/stanmoore1/lammps/pull/16 with some cleanups and a HIP bugfix, which would merge into this PR branch.

**Comment by skyreflectedinmirrors** [comment]
Can confirm, this builds w/ the latest kokkos develop on MI-2XX

**Comment by stanmoore1** [comment]
@akohlmey would like to get this merged ASAP now that the stable is out.

**Comment by akohlmey** [comment]
> @akohlmey would like to get this merged ASAP now that the stable is out.

Yup. I am working through the back log right now.

**Pull request merged**
